vector/src/androidTest/java/im/vector/app/ui/robot/SpaceRobot.kt

```diff
@@ -0,0 +1,214 @@
+import im.vector.app.features.spaces.leave.SpaceLeaveAdvancedActivity
+import im.vector.app.features.spaces.manage.SpaceManageActivity
+import org.hamcrest.Matchers.allOf
+import java.util.UUID
+
+class SpaceRobot {
+
+    fun createSpace() {
+        clickOn(R.string.add_space)
+        waitUntilActivityVisible<SpaceCreationActivity> {
+            waitUntilViewVisible(withId(R.id.privateButton))
+        }
+        crawlCreate()
+
+        onView(withId(R.id.roomList))
+                .perform(
+                        RecyclerViewActions.actionOnItem<RecyclerView.ViewHolder>(
+                                ViewMatchers.hasDescendant(withText(R.string.room_displayname_empty_room)),
+                                click()
+                        ).atPosition(0)
+                )
+        clickOn(R.id.spaceAddRoomSaveItem)
+        waitUntilActivityVisible<HomeActivity> {
+            waitUntilViewVisible(withId(R.id.roomListContainer))
+        }
+    }
+
+    private fun crawlCreate() {
+        //public
+        clickOn(R.id.publicButton)
+        waitUntilViewVisible(withId(R.id.recyclerView))
+        onView(withHint(R.string.create_room_name_hint)).perform(replaceText(UUID.randomUUID().toString()))
+        clickOn(R.id.nextButton)
+        waitUntilViewVisible(withId(R.id.recyclerView))
+        pressBack()
+        pressBack()
+
+        //private
+        clickOn(R.id.privateButton)
+        waitUntilViewVisible(withId(R.id.recyclerView))
+        clickOn(R.id.nextButton)
+
+        waitUntilViewVisible(withId(R.id.teammatesButton))
+        //me and teammates
+        clickOn(R.id.teammatesButton)
+        waitUntilViewVisible(withId(R.id.recyclerView))
+        clickOn(R.id.nextButton)
+        pressBack()
+        pressBack()
+
+        //just me
+        waitUntilViewVisible(withId(R.id.justMeButton))
+        clickOn(R.id.justMeButton)
+        waitUntilActivityVisible<SpaceManageActivity> {
+            waitUntilViewVisible(withId(R.id.roomList))
+        }
+    }
+
+    fun openSpaceMenu() {
+        waitUntilViewVisible(withId(R.id.groupListView))
+        onView(withId(R.id.groupListView))
+                .perform(
+                        RecyclerViewActions.actionOnItem<RecyclerView.ViewHolder>(
+                                ViewMatchers.hasDescendant(allOf(withId(R.id.groupTmpLeave), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE))),
+                                clickChildWithId(R.id.groupTmpLeave)
+                        ).atPosition(0)
+                )
+        waitUntilDialogVisible(withId(R.id.spaceNameView))
+    }
+
+    fun invitePeople() {
+        clickOn(R.id.invitePeople)
+        waitUntilDialogVisible(withId(R.id.inviteByMxidButton))
+        clickOn(R.id.inviteByMxidButton)
+        waitUntilActivityVisible<InviteUsersToRoomActivity> {
+            waitUntilViewVisible(withId(R.id.userListRecyclerView))
+        }
+        EspressoHelper.getCurrentActivity()!!.finish()
```

Comment:
in this case it will require to call it two times to close activity and also there will be the third one few lines below and it looks a little bit weird, isn't it better to just close activity explicitly, since what's what we want to achieve?

Reply:
you're correct and both methods achieve the same result, the main difference is that calling finish directly is not possible for real users so there's the potential that our tests end up with different behaviour

I agree that adding all the _backs_ is a bit tedious and verbose but it does provide interesting UX commentary 

```
// close keyboard
pressBack()

// close popup
pressBack()

// exit screen
pressBack()
```